RATIS-873. Fix Install Snapshot Notification

hanishakoneru opened the pull request:
This Jira aims to fix the following:

- When Follower is in the process of installing snapshot and it gets an append entry, it replies with result INCONSISTENCY and the follower next index is updated to the snapshot index being installed. This should not happen as the snapshot installation is still in progress. Follower's next index on the leader should remain the same.

- After InstallSnapshot is done, Leader should update the commitIndex of the Follower to the installed snapshot index.

- After InstallSnapshot, when reloading StateMachine, any previously open LogSegment should be closed, if the last entry in the open log is already included in the snapshot.

- When Follower is notified to install snapshot through StateMachine, the reply should indicate the same. It would help with debugging if the Install Snapshot success reply and Install Snapshot notified replies are distinct.

hanishakoneru commented:
Failing unit tests pass locally.

lokeshj1703 commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -369,13 +369,22 @@ public void onNext(InstallSnapshotReplyProto reply) {
-          LOG.info("{}: set follower snapshotIndex to {}.", this, followerSnapshotIndex);
+          LOG.info("{}: Already Installed Snapshot Index {}.", this, followerSnapshotIndex);
           getFollower().setSnapshotIndex(followerSnapshotIndex);
+          updateCommitIndex(followerSnapshotIndex);
```
Should we also update nextIndex?

hanishakoneru replied:
Done.

lokeshj1703 commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -394,17 +403,19 @@ public void onNext(InstallSnapshotReplyProto reply) {
     @Override
     public void onCompleted() {
-      LOG.info("{}: follower responses installSnapshot COMPLETED", this);
+      if (!firstResponseReceived) {
```
Why do we need this check?

hanishakoneru replied:
I had added it as otherwise the Log is flooded with "InstallSnapshot COMPLETED" messages even when the install snapshot is in progress. I updated it to not print in case of install snapshot notification.

lokeshj1703 commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -394,17 +403,19 @@ public void onNext(InstallSnapshotReplyProto reply) {
+      GrpcUtil.warn(LOG, () -> this + ": Failed InstallSnapshot", t);
+      grpcServerMetrics.onRequestRetry(); // Update try counter
       resetClient(null);
-      close();
```
We can add a flag in close to mark it as closed without notifying append in this case.

hanishakoneru replied:
Done.

hanishakoneru commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -369,13 +369,22 @@ public void onNext(InstallSnapshotReplyProto reply) {
 
       switch (reply.getResult()) {
         case SUCCESS:
+          LOG.info("{}: Completed InstallSnapshot. Reply: {}", this);
```
Done.

hanishakoneru commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -369,13 +369,22 @@ public void onNext(InstallSnapshotReplyProto reply) {
+          LOG.info("{}: Completed InstallSnapshot. Reply: {}", this);
+          removePending(reply);
+          break;
+        case NOTIFIED:
```
Agree. Done.

hanishakoneru commented:
Thank you @lokeshj1703 for the review. I addressed all your comments in commit [8ea4014](https://github.com/apache/incubator-ratis/pull/88/commits/8ea401453866d0d9e5a2b099fc6b8ecd6bdee33a).

lokeshj1703 commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -222,6 +222,10 @@ private void increaseNextIndex(AppendEntriesRequestProto request) {
   }
 
+  private void increaseNextIndex(final long installedSnapshotIndex) {
+    getFollower().increaseNextIndex(installedSnapshotIndex + 1);
```
FollowerInfo#increaseNextIndex will throw error if current next index is greater than installedSnapshotIndex + 1. I think we should do something like updateCommitIndex here.

hanishakoneru replied:
Done.

lokeshj1703 commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -369,13 +381,19 @@ public void onNext(InstallSnapshotReplyProto reply) {
 
       switch (reply.getResult()) {
         case SUCCESS:
```
Do we need to update commitIndex and nextIndex here as well?

hanishakoneru replied:
Yes but I am not sure what all needs to be fixed in the regular InstallSnapshot. I can update the snapshotIndex, commitIndex and nextIndex here for now.

lokeshj1703 replied:
Yes..Makes sense to do it in a separate jira.

lokeshj1703 commented on ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:
```diff
@@ -394,17 +412,24 @@ public void onNext(InstallSnapshotReplyProto reply) {
     @Override
     public void onCompleted() {
-      LOG.info("{}: follower responses installSnapshot COMPLETED", this);
+      if (!isNotificationOnly) {
+        LOG.info("{}: follower responded installSnapshot COMPLETED", this);
+      } else {
+        if (LOG.isDebugEnabled()) {
+          LOG.info("{}: follower responded installSnapshot COMPLETED", this);
+        }
+      }
```
We can combine the if statements using below condition.
`!isNotificationOnly || LOG.isDebugEnabled()`

hanishakoneru replied:
Done.

lokeshj1703 commented on ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java:
```diff
@@ -492,6 +492,13 @@ public void syncWithSnapshot(long lastSnapshotIndex) {
+    if (openSegment != null && openSegment.getEndIndex() <= lastSnapshotIndex) {
+      fileLogWorker.closeLogSegment(openSegment);
+    }
+    cache.clear();
```
We can potentially clear closed log segments with endIndex > snapshot index? Maybe we should move it inside the if statement?

hanishakoneru replied:
Done.

hanishakoneru commented:
Thanks @lokeshj1703 for the reviews. Merged into master.
